src/robot/running/arguments/typeconverters.py

```diff
@@ -389,61 +389,165 @@ class ListConverter(TypeConverter):
-        return self._literal_eval(value, list)
+    _convert = _non_string_convert
 
 
 @TypeConverter.register
 class TupleConverter(TypeConverter):
     type = tuple
     type_name = 'tuple'
     value_types = (str, Sequence)
+    aliases = ('tuple',)
 
-    def _non_string_convert(self, value, explicit_type=True):
-        return tuple(value)
-
-    def _convert(self, value, explicit_type=True):
-        return self._literal_eval(value, tuple)
+    def __init__(self, type_, custom_converters=None, languages=None):
+        super().__init__(type_, languages=languages)
+        self.nested_types = getattr(type_, '__args__', ())
 
+    def _non_string_convert(self, value, explicit_type=True):
+        if isinstance(value, str):
+            converted_tuple = self._literal_eval(value, tuple)
+        elif is_list_like(value):
+            converted_tuple = tuple(value)
+        else:
+            raise ValueError
+        if self.nested_types:
+            tmp_list = list(converted_tuple)
+            for i, elem in enumerate(converted_tuple):
+                converter = TypeConverter.converter_for(self.nested_types[i],
+                                                        languages=self.languages)
+                if converter:
+                    types = ', '.join([type_name(t) for t in self.nested_types])
+                    tmp_list[i] = converter.convert(f"Tuple[{types}]", elem,
+                                                    explicit_type)
+            converted_tuple = tuple(tmp_list)
+        return converted_tuple
+
+    _convert = _non_string_convert
 
 @TypeConverter.register
 class DictionaryConverter(TypeConverter):
     type = dict
     abc = Mapping
+    abc = (Mapping, MutableMapping)
     type_name = 'dictionary'
-    aliases = ('dict', 'map')
+    aliases = ('dict', 'dictionary', 'map')
     value_types = (str, Mapping)
 
-    def _non_string_convert(self, value, explicit_type=True):
-        if issubclass(self.used_type, dict) and not isinstance(value, dict):
-            return dict(value)
-        return value
+    def __init__(self, type_, custom_converters=None, languages=None):
+        super().__init__(type_, languages=languages)
+        self.nested_types = getattr(type_, '__args__', ())
 
-    def _convert(self, value, explicit_type=True):
-        return self._literal_eval(value, dict)
+    def no_conversion_needed(self, value):
+        return self.used_type in self.abc and issubclass(type(value), self.abc)
+
+    def _non_string_convert(self, value, explicit_type=True):
+        if isinstance(value, str):
+            converted_dict = self._literal_eval(value, dict)
+        elif is_dict_like(value):
+            converted_dict = dict(value)
+        else:
+            raise ValueError
+        if self.nested_types and len(self.nested_types) == 2:
+            for key, elem in converted_dict.copy().items():
+                key_type = self.nested_types[0]
+                value_type = self.nested_types[1]
```

Comment:
Having
```python
key_type, value_type = self.nested_types
```
before the loop would be better. Possibly that assignment could be done already in `__init__` and then `self.nested_types` wouldn't be needed at all. Other converters should obviously use the same logic in that case.